File: apps/web/src/types/release.ts
import type { PaginatedResponse, PaginationParams } from "./pagination";
import type { SummaryItem } from "./summary";

type ReleaseStatus = "Planned" | "QA" | "Staged" | "Shipped" | "Canceled";

type ReleaseType = "Major" | "Minor" | "Patch" | "Hotfix";

type Release = {
  id: string;
  name: string;
  version: string;
  status: ReleaseStatus;
  releaseType: ReleaseType;
  owner: string;
  targetDate: string;
  shippedAt: string | null;
  summary: string;
  internalNotes: string;
  publicNotes: string;
  includedFeatureIds: string[];
  includedBugIds: string[];
  isPublic: boolean;
};

type ReleaseCreateInput = Omit<Release, "id">;

type ReleaseUpdateInput = Partial<ReleaseCreateInput>;

type ReleaseFilters = {
  search: string;
  status?: ReleaseStatus;
  releaseType?: ReleaseType;
  isPublic?: boolean;
};

type ReleaseListResponse = PaginatedResponse<Release>;

type ReleaseSummary = SummaryItem;

export type {
  PaginationParams,
  Release,
  ReleaseCreateInput,
  ReleaseFilters,
  ReleaseListResponse,
  ReleaseStatus,
  ReleaseSummary,
  ReleaseType,
  ReleaseUpdateInput,
};

File: apps/api/app/schemas/release.py
from typing import Literal

from pydantic import BaseModel

from app.schemas.pagination import PaginatedResponse


ReleaseStatus = Literal["Planned", "QA", "Staged", "Shipped", "Canceled"]
ReleaseType = Literal["Major", "Minor", "Patch", "Hotfix"]


class ReleaseCreate(BaseModel):
    name: str
    version: str
    status: ReleaseStatus
    release_type: ReleaseType
    owner: str
    target_date: str
    shipped_at: str | None = None
    summary: str
    internal_notes: str
    public_notes: str
    included_feature_ids: list[str]
    included_bug_ids: list[str]
    is_public: bool


class ReleaseUpdate(BaseModel):
    name: str | None = None
    version: str | None = None
    status: ReleaseStatus | None = None
    release_type: ReleaseType | None = None
    owner: str | None = None
    target_date: str | None = None
    shipped_at: str | None = None
    summary: str | None = None
    internal_notes: str | None = None
    public_notes: str | None = None
    included_feature_ids: list[str] | None = None
    included_bug_ids: list[str] | None = None
    is_public: bool | None = None


class ReleaseItem(BaseModel):
    id: str
    name: str
    version: str
    status: ReleaseStatus
    release_type: ReleaseType
    owner: str
    target_date: str
    shipped_at: str | None
    summary: str
    internal_notes: str
    public_notes: str
    included_feature_ids: list[str]
    included_bug_ids: list[str]
    is_public: bool

    model_config = {"from_attributes": True}


ReleaseListResponse = PaginatedResponse[ReleaseItem]